pkg/controller/pxc/users.go

```diff
@@ -0,0 +1,284 @@
+package pxc
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+
+	api "github.com/percona/percona-xtradb-cluster-operator/pkg/apis/pxc/v1"
+	"github.com/percona/percona-xtradb-cluster-operator/pkg/pxc/users"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const internalPrefix = "internal-"
+
+func (r *ReconcilePerconaXtraDBCluster) reconcileUsers(cr *api.PerconaXtraDBCluster) error {
+	if cr.Status.Status != api.AppStateReady {
+		return nil
+	}
+
+	err := r.handleSysUsersSecret(cr)
+	if err != nil {
+		log.Error(err, "handle system users secret")
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) handleSysUsersSecret(cr *api.PerconaXtraDBCluster) error {
+	sysUsersSecretObj := corev1.Secret{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      cr.Spec.SecretsName,
+		},
+		&sysUsersSecretObj,
+	)
+	if err != nil && k8serrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return errors.Wrapf(err, "get sys users secret '%s'", cr.Spec.SecretsName)
+	}
+
+	newSysData, err := json.Marshal(sysUsersSecretObj.Data)
+	if err != nil {
+		return errors.Wrap(err, "marshal sys secret data")
+	}
+	newSecretDataHash := sha256Hash(newSysData)
+
+	internalSysSecretObj, err := r.getInternalSysUsersSecret(cr, &sysUsersSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "get internal sys users secret")
+	}
+
+	dataChanged, err := sysUsersSecretDataChanged(newSecretDataHash, &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "check sys users data changes")
+	}
+
+	if !dataChanged {
+		return nil
+	}
+
+	var restartPXC bool
+	var restartProxy bool
+
+	err = r.manageSysUsers(cr, &sysUsersSecretObj, &internalSysSecretObj, &restartPXC, &restartProxy)
+	if err != nil {
+		return errors.Wrap(err, "manage sys users")
+	}
+
+	err = r.updateInternalSysUsersSecret(cr, &sysUsersSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "update internal sys users secret")
+	}
+
+	if restartProxy && cr.Spec.ProxySQL != nil && cr.Spec.ProxySQL.Size > 0 {
+		err = r.restartProxy(cr, newSecretDataHash)
+		if err != nil {
+			return errors.Wrap(err, "restart proxy")
+		}
+	}
+
+	if restartPXC {
+		err = r.restartPXC(cr, newSecretDataHash, &sysUsersSecretObj)
+		if err != nil {
+			return errors.Wrap(err, "restart pxc")
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) manageSysUsers(cr *api.PerconaXtraDBCluster, sysUsersSecretObj, internalSysSecretObj *corev1.Secret, restartPXC, restartProxy *bool) error {
```

Comment:
I think `restartPXC`, `restartProxy` should not be parameters, but returned result.

Reply:
done https://github.com/percona/percona-xtradb-cluster-operator/pull/418/commits/9a770955143b4e0e1b00e53653fdb45ab333357e